src/HexagonOptimize.cpp

```diff
@@ -0,0 +1,1084 @@
+    }
+
+    template <typename NodeType, typename LetType>
+    void visit_let(NodeType &result, const LetType *op) {
+        Expr value = mutate(op->value);
+        string deinterleaved_name = op->name + ".deinterleaved";
+        NodeType body;
+        if (is_native_interleave(value)) {
+            // We can provide a deinterleaved version of this let value.
+            vars.push(deinterleaved_name, true);
+            body = mutate(op->body);
+            vars.pop(deinterleaved_name);
+        } else {
+            body = mutate(op->body);
+        }
+        if (value.same_as(op->value) && body.same_as(op->body)) {
+            result = op;
+        } else if (body.same_as(op->body)) {
+            // If the body didn't change, we must not have used the deinterleaved value.
+            result = LetType::make(op->name, value, body);
+        } else {
+            // We need to rewrap the body with new lets.
+            result = body;
+            bool deinterleaved_used = uses_var(result, deinterleaved_name);
+            bool interleaved_used = uses_var(result, op->name);
+            if (deinterleaved_used && interleaved_used) {
+                // The body uses both the interleaved and
+                // deinterleaved version of this let. Generate both
+                // lets, using the deinterleaved one to generate the
+                // interleaved one.
+                Expr deinterleaved = remove_interleave(value);
+                Expr deinterleaved_var = Variable::make(deinterleaved.type(), deinterleaved_name);
+                result = LetType::make(op->name, native_interleave(deinterleaved_var), result);
+                result = LetType::make(deinterleaved_name, deinterleaved, result);
+            } else if (deinterleaved_used) {
+                // Only the deinterleaved value is used, we can eliminate the interleave.
+                result = LetType::make(deinterleaved_name, remove_interleave(value), result);
+            } else if (interleaved_used) {
+                // Only the original value is used, regenerate the let.
+                result = LetType::make(op->name, value, result);
+            } else {
+                // The let must have been dead.
+                internal_assert(!uses_var(op->body, op->name)) << "EliminateInterleaves eliminated a non-dead let.\n";
+            }
+        }
+    }
+
+    void visit(const Let *op) { visit_let(expr, op); }
+    void visit(const LetStmt *op) { visit_let(stmt, op); }
+
+    void visit(const Cast *op) {
+        if (op->type.bits() == op->value.type().bits()) {
+            // We can only move interleaves through casts of the same size.
+
+            Expr value;
+            bool interleave;
+            std::tie(value, interleave) = mutate_with_interleave(op->value);
+
+            if (interleave) {
+                expr = native_interleave(Cast::make(op->type, value));
+            } else if (is_native_interleave(value)) {
```

Comment:
This case shouldn't be needed: if it was an interleave, mutate_with_interleave should have indicated such.